### ISSUE TITLE
[Feature request]: Support various x64 architecture level (gcc's x86-64, ..., x86-64-v4)

### ISSUE THREAD

**Issue by abique** [flatpak/flatpak#4621]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

Hi,
I think there is an opportunity to achieve higher performances with flatpak applications.
What if instead of simply "x86-64" architecture for intel we would also support "x86-64-vX" as well?

As a reference, gcc has an option `-march=...` which can select the instruction set to compile for: https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

The current architecture can be auto-detected at runtime and we can start optimized applications and take advantages of all the CPU features instead of running compatible ones (which are built for first generation of x86-64).

I suppose this will lead to:
- higher performances
- maybe lower power consumption
- more work for the buildbot, but I believe that the additional compilation time is acceptable if we consider the wins on the user side
- slightly more complex infrastructure but in the end it should be equivalent to running an additional architecture

What do you think?
Regards,
Alex

**Comment by Erick555** [comment]
I think this feature request is for flatpak runtimes rather than flatpak itself.

**Comment by abique** [comment]
> I think this feature request is for flatpak runtimes rather than flatpak itself. Flatpak will install & run whatever runtime contains.

Indeed this seems correct, but I believe it needs also a bit of thought on the flatpak side isn't it?

**Comment by mwleeds** [comment]
There's discussion in #4594 on what Flatpak would need to do to support more architectures.

**Comment by smcv** [comment]
> There's discussion in #4594 on what Flatpak would need to do to support more architectures.

I think it's important to be clear about whether we are talking about:

* interoperable ABI families
    * e.g. i386, i486, i586 and i686 are all one family, which we call i386
    * e.g. base x86_64 up to x86_64-v4 are a different family, which we call x86_64
    * if you have a library compiled for i486 and a library compiled for i686, it's completely OK to load them into the same process space and have them call into each other, so i486 and i686 are the same family
    * (but if you have an ARM softfloat library and an ARM hardfloat library, it is *not* OK to load them into the same process space and have them call into each other, so ARM softfloat and ARM hardfloat are different families)

* the individual members of those families, such as i486

because people call both of those things, interchangeably, "architectures".

**Comment by abique** [comment]
In our case, x86_64-vN+1 is a superset of x86_64-vN.
So yes anything compiled for vN works with a vN+1 processor. But something compiled for vN+1 won't work with a vN processor.

**Comment by yhm3ps1jfp76nkd** [comment]
Hi, any update? Now there is ARM64 for many apps on FlatHub already

**Comment by ispanos** [comment]
I found this issue after listening to this [interview](https://www.youtube.com/watch?v=uBNY0hxCfG8) of Matthew Miller (Fedora Project leader) (@1:09:50).

@Erick555

> I think this feature request is for flatpak runtimes rather than flatpak itself.

Does this mean that I can ask a publisher to create an x86_64-v4 binary and `flatpak install` with choose that over  x86_64?

**Comment by smcv** [comment]
> Does this mean that I can ask a publisher to create an x86_64-v4 binary and `flatpak install` with choose that over x86_64?

No, Flatpak has no concept of architecture levels, only architecture families. A repository like Flathub can offer different builds of the same app or runtime for different architecture families (like x86_64, i386 and arm64), but it cannot offer different builds of the same app or runtime for different CPU functionality levels (like x86_64-v1 and x86_64-v4, or i486 and i686).

If an app publisher builds their app for x86_64-v4, then that's their choice to make, but if they do that, then their app will not work on CPUs of a lower architecture level than the baseline they have chosen. It is not possible to publish both x86_64-v1 and x86_64-v4 and have Flatpak automatically choose the newest one supported by the CPU it is currently running on. That's why this feature request remains open: if the feature was implemented, then this feature request would have been closed.

(The equivalent feature is also not available in many traditional package managers, like Debian's apt: you can install `foo:amd64` or `foo:i386` or `foo:arm64`, which are different architecture families, but a single apt repository can't offer separate x86_64-v4 and x86_64-v1 builds of foo, or separate i486 and i686 builds of foo.)

**Comment by ispanos** [comment]
Thanks for clarifying. Its a feature I would very much like to see implemented and I had to ask.

**Comment by mainrs** [comment]
> No, Flatpak has no concept of architecture levels, only architecture families

What is the stance on changing this? Would there be support in doing so when appropriate discussions, implementation details and finally an implementation have been provided? What would be the right channels to start such a discussion?
